core/src/main/scala/kafka/log/Log.scala

```diff
@@ -565,8 +565,8 @@ class Log(val dir: File,
   def deleteOldSegments(predicate: LogSegment => Boolean): Int = {
     // find any segments that match the user-supplied predicate UNLESS it is the final segment
     // and it is empty (since we would just end up re-creating it
-    val lastSegment = activeSegment
-    val deletable = logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastSegment.baseOffset || s.size > 0))
+    val lastEntry = segments.lastEntry
+    val deletable = logSegments.takeWhile(s => predicate(s) && (lastEntry != null && s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
```

Comment:
Thanks Jun, sounds good. Sorry for leading you in the wrong direction @omkreddy, we understand the motivation better now. :)